apps/admin.py

```diff
@@ -47,3 +69,72 @@ def field_func(self, obj):
         return field_func
 
     return wrap
+
+
+class SafeJson(str):
+    def __html__(self):
+        return self
+
+
+def pretty_json_print(field, indent=2, sort_keys=True, line_separator=u'\n'):
+    response = json.dumps(field, sort_keys=sort_keys, indent=indent)
+
+    # Get the Pygments formatter
+    formatter = HtmlFormatter(style='colorful', lineseparator=line_separator)
+
+    # Highlight the data
+    response = highlight(response, JsonLexer(), formatter)
+
+    # Get the stylesheet
+    style = "<style>" + formatter.get_style_defs() + "</style><br>"
+
+    # Safe the output
+    return SafeJson(style + response)
+
+
+def object_detail_admin_link(obj):
+    """
+    Returns the url admin link of the object passed in argument
+    """
+    return format_html(
+        '<a href="{}">{}</a>',
+        admin_change_url(obj),
+        obj.id if obj else ''
+    )
+
+
+class ShipmentAdminDisplayMixin:
+    """
+    This mixin aims at facilitating the addition of a shipment link on a list
+    or detail page.
+
+    example:
+        class MyModelAdmin(ShipmentAdminDisplayMixin, ModelAdmin):
+            ...
+
+            list_display = (
+                ...
+                'shipment_display',
+            )
+
+            # Or for detail display
+            readonly_fields = (
+                ...
+                'shipment_display',
+            )
+    """
+    def shipment_display(self, obj):
+        return object_detail_admin_link(obj.shipment)
+
+    shipment_display.short_description = "Shipment"
+
+
+class NoAddUpdateDeletePermissionMixin:
```

Comment:
Maybe rename this to "NoWritePermissionMixin"?

Reply:
Or maybe ReadOnlyPermissionMixin

Reply:
Sounds better thanks!